### DISCUSSION TITLE
Distributed cache

### DISCUSSION THREAD

**Discussion by safa-a-safa** [keycloak/keycloak#19674] [Q&A]
Hi,

#### **Version: Keycloak 21.0.2**

**TLDR:**
When I have two or more keycloaks running I would like them to sync user and permissions, by default they are local cache.
also not sure if the local cache syncs at any time as I can't find any documentation on that.

Im having an issue with the caching specifically realms and users, i am using this page from the docs: https://www.keycloak.org/server/caching

When I have two keycloaks running the user permissions are not syncing between the two of them. I have tried to create a custom cache config and I have switched **realms** and **users** to **distributed-cache**.

in the docker file I am adding the file to the **/opt/keycloak/conf/** location and setting **ENV KC_CACHE_CONFIG_FILE="custom-cache.xml"** and **ENV KC_CACHE=ispn**

This is my custom file, it is just a copy of the default with the changes made to **realms** and **users**:

```xml
<?xml version="1.0" encoding="UTF-8"?>

<infinispan
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="urn:infinispan:config:11.0 http://www.infinispan.org/schemas/infinispan-config-11.0.xsd"
    xmlns="urn:infinispan:config:11.0">

    <cache-container name="keycloak">
        <transport lock-timeout="60000"/>
        <distributed-cache name="realms" owners="2">
            <encoding>
                <key media-type="application/x-java-object" />
                <value media-type="application/x-java-object" />
            </encoding>
            <expiration lifespan="-1" />
        </distributed-cache>
        <distributed-cache name="users" owners="2">
            <encoding>
                <key media-type="application/x-java-object" />
                <value media-type="application/x-java-object" />
            </encoding>
            <expiration lifespan="-1" />
        </distributed-cache>
        <distributed-cache name="sessions" owners="2">
            <expiration lifespan="-1" />
        </distributed-cache>
        <distributed-cache name="authenticationSessions" owners="2">
            <expiration lifespan="-1" />
        </distributed-cache>
        <distributed-cache name="offlineSessions" owners="2">
            <expiration lifespan="-1" />
        </distributed-cache>
        <distributed-cache name="clientSessions" owners="2">
            <expiration lifespan="-1" />
        </distributed-cache>
        <distributed-cache name="offlineClientSessions" owners="2">
            <expiration lifespan="-1" />
        </distributed-cache>
        <distributed-cache name="loginFailures" owners="2">
            <expiration lifespan="-1" />
        </distributed-cache>
        <local-cache name="authorization">
            <encoding>
                <key media-type="application/x-java-object" />
                <value media-type="application/x-java-object" />
            </encoding>
            <memory max-count="10000" />
        </local-cache>
        <replicated-cache name="work">
            <expiration lifespan="-1" />
        </replicated-cache>
        <local-cache name="keys">
            <encoding>
                <key media-type="application/x-java-object" />
                <value media-type="application/x-java-object" />
            </encoding>
            <expiration max-idle="3600000" />
            <memory max-count="1000" />
        </local-cache>
        <distributed-cache name="actionTokens" owners="2">
            <encoding>
                <key media-type="application/x-java-object" />
                <value media-type="application/x-java-object" />
            </encoding>
            <expiration max-idle="-1" lifespan="-1" interval="300000" />
            <memory max-count="-1" />
        </distributed-cache>
    </cache-container>
</infinispan>
```

When I start the containers I am unable to tell if it is using my new file.
I can't seem to figure out what I am doing wrong, Is there an ENV I'm missing? or am I going about this the wrong way.

Any help with this would be much appreciated.
